opensearch/src/main/java/org/opensearch/sql/opensearch/storage/OpenSearchIndexScan.java

```diff
@@ -73,118 +74,56 @@ public OpenSearchIndexScan(OpenSearchClient client,
       Settings settings, OpenSearchRequest.IndexName indexName,
       OpenSearchExprValueFactory exprValueFactory) {
     this.client = client;
-    this.request = new OpenSearchQueryRequest(indexName,
-        settings.getSettingValue(Settings.Key.QUERY_SIZE_LIMIT), exprValueFactory);
+    this.requestBuilder = new OpenSearchRequestBuilder(indexName, settings,
+        new SearchSourceBuilder(), exprValueFactory);
   }
 
   @Override
   public void open() {
     super.open();
-
-    // For now pull all results immediately once open
-    List<OpenSearchResponse> responses = new ArrayList<>();
-    OpenSearchResponse response = client.search(request);
-    while (!response.isEmpty()) {
-      responses.add(response);
-      response = client.search(request);
-    }
-    iterator = Iterables.concat(responses.toArray(new OpenSearchResponse[0])).iterator();
+    querySize = requestBuilder.getSourceBuilder().size();
+    isAggregation = !(null == requestBuilder.getSourceBuilder().aggregations());
+    request = requestBuilder.build();
+    iterator = Collections.emptyIterator();
+    queryCount = 0;
+    fetchNextBatch();
   }
 
   @Override
   public boolean hasNext() {
+    if (isAggregation) {
```

Comment:
aggregation should also support scroll, right?

Reply:
yes. I'll work on it next. Aggregation works quite different from non-agg queries. If there's a lot to change I'll consider making a separate PR for it.

Reply:
could open a issue if it is not covered in this PR?